controllers/operatorpolicy_status.go

```diff
@@ -0,0 +1,429 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	policyv1 "open-cluster-management.io/config-policy-controller/api/v1"
+	policyv1beta1 "open-cluster-management.io/config-policy-controller/api/v1beta1"
+)
+
+// updateStatus takes one condition to update, and related objects for that condition. The related
+// objects given will replace all existing relatedObjects with the same gvk. If a condition is
+// changed, the compliance will be recalculated and a compliance event will be emitted. The
+// condition and related objects can match what is already in the status - in that case, no API
+// calls are made. The `lastTransitionTime` on a condition is not considered when checking if the
+// condition has changed. If not provided, the `lastTransitionTime` will use "now". It also handles
+// preserving the `CreatedByPolicy` property on relatedObjects.
+//
+// This function assumes that all given related objects are of the same kind.
+//
+// Note that only changing the related objects will not emit a new compliance event, but will update
+// the status.
+func (r *OperatorPolicyReconciler) updateStatus(
+	ctx context.Context,
+	policy *policyv1beta1.OperatorPolicy,
+	updatedCondition metav1.Condition,
+	updatedRelatedObjs ...policyv1.RelatedObject,
+) error {
+	condChanged := false
+
+	if updatedCondition.LastTransitionTime.IsZero() {
+		updatedCondition.LastTransitionTime = metav1.Now()
+	}
+
+	condIdx, existingCondition := policy.Status.GetCondition(updatedCondition.Type)
+	if condIdx == -1 {
+		condChanged = true
+
+		// Just append, the conditions will be sorted later.
+		policy.Status.Conditions = append(policy.Status.Conditions, updatedCondition)
+	} else if conditionChanged(updatedCondition, existingCondition) {
+		condChanged = true
+
+		policy.Status.Conditions[condIdx] = updatedCondition
+	}
+
+	if condChanged {
+		updatedComplianceCondition := calculateComplianceCondition(policy)
+
+		compCondIdx, _ := policy.Status.GetCondition(updatedComplianceCondition.Type)
```

Comment:
Is this different than `condIdx`?

Reply:
Yes, `compCondIdx` is the index of the (overall) Compliance condition; `condIdx` is the index of the more specific condition that was passed in to be updated. When the passed-in condition does not match what is currently in the status, the overall compliance needs to be recalculated and possibly updated.